### PR TITLE
Add the UCI Adult dataset

### LINKED ISSUE #169: [Benchmarks] Add the UCI Adult dataset
Add the Adult dataset from the UCI Machine Learning Repository as a benchmark dataset.

### PR DESCRIPTION
**Pull request by A-Salmon**
This pull request will add the 'Adult' dataset form the UCI machine learning repository to the package. In more detail, it includes the following:

- Implemented loader functions with associated documentation in `src/data/tabular.jl`.
- Exported the functions in `Data.jl`
- Updated the `tabular.qmd` notebook in the artifacts
- Updated the `students_documentation/summary.qmd` file to reflect the additions
- Added the dataset to the release artifacts
- Added testing for the dataset

Closes #169

### CODE REVIEW

**Comment by VincentPikand** [comment]
your PR is to main

File: src/data/tabular/adult.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Load and preprocesses data from the UCI 'Adult' dataset
+
+# Arguments
+- `n::Union{Nothing, Int}=nothing`: The number of samples to subsample from the dataset.
+
+# Returns
+- `counterfactual_data::CounterfactualData`: A `CounterfactualData` object containing the preprocessed data.
+
+# Example
+data = load_german_credit(20) # loads and preprocesses 20 samples from the Adult dataset
```
Comment (VincentPikand) [inline src/data/tabular/adult.jl:13]: comments not updated

File: src/data/tabular/adult.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+"""
+function load_uci_adult(n::Union{Nothing,Int}=1000)
+    # Throw an exception if n > 1000:
+    if !isnothing(n) && n > 1000
+        throw(ArgumentError("n must be <= 1000"))
+    end
```
Comment (VincentPikand) [inline src/data/tabular/adult.jl:19]: I think this only applies to german creidt, but a check of ` n <= 0` should still be here

File: src/data/tabular/adult.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            :hours_per_week,
+            :native_country,
+            :target,
+        ],
+    )
+
+    # Preprocessing
+    transformer = Standardizer(; count=true)
+    mach = MLJBase.fit!(machine(transformer, df[:, Not(:target)]))
+    X = MLJBase.transform(mach, df[:, Not(:target)])
```
Comment (VincentPikand) [inline src/data/tabular/adult.jl:51]: `Not` is from `DataFrames, change all imports from `using` to `import` and then see if it still compiles, update everything and then change back

File: students_documentation/summary_notebook/summary.qmd [hunk 4]
```diff
@@ -249,6 +249,44 @@ end
...
+```{julia}
+n_individuals = 5
+
+# Choose random factual individuals
+ids = rand(findall(predict_label(M, german_credit_data) .== factual), n_individuals)
+xs = select_factual(german_credit_data, ids)
+
+generator = GenericGenerator()
+
+generated_counterfactuals_german_credit_data = generate_counterfactual(
```
Comment (VincentPikand) [inline students_documentation/summary_notebook/summary.qmd:280]: german credit here, go over all of these please

File: students_documentation/summary_notebook/summary.qmd [hunk 5]
```diff
@@ -249,6 +249,44 @@ end
...
+    xs,
+    target,
+    german_credit_data,
```
Comment (VincentPikand) [inline students_documentation/summary_notebook/summary.qmd:283]: .

File: src/data/tabular/adult.jl (hunk 1 above)
Comment (A-Salmon) [reply src/data/tabular/adult.jl:13]: Right, good catch

File: src/data/tabular/adult.jl (hunk 2 above)
Comment (A-Salmon) [reply src/data/tabular/adult.jl:19]: Alright that makes sense, will change this.

File: students_documentation/summary_notebook/summary.qmd (hunk 4 above)
Comment (A-Salmon) [reply students_documentation/summary_notebook/summary.qmd:280]: This was very sloppy of me, I'll pay better attention next time.

File: src/data/tabular/adult.jl (hunk 3 above)
Comment (A-Salmon) [reply src/data/tabular/adult.jl:51]: Ahh I see, good way of making these calls more explicit. I'll fix this, thanks!

**Review by VincentPikand — APPROVED** [review]
LGTM. A quick tip: use VSCode's or any other editors git abilities to view all of your changes before committing, this way you can remove any unnecessary `println`'s or wrong comments or whatever. Generally great for checking your work in any project.

Screenshot 2023-06-16 at 15 19 23

File: src/data/tabular/adult.jl [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        ],
+    )
+
+    # Preprocessing
+    transformer = Standardizer(; count=true)
+    mach = MLJBase.fit!(machine(transformer, df[:, DataFrames.Not(:target)]))
+    X = MLJBase.transform(mach, df[:, DataFrames.Not(:target)])
+    X = Matrix(X)
+    X = permutedims(X)
+    println(typeof(X))
```
Comment (RaunoArike) [inline src/data/tabular/adult.jl:50]: Please remove the print statement from here.

File: test/data.jl [hunk 7]
```diff
@@ -61,3 +61,16 @@ end
...
+    counterfactual_data = load_uci_adult()
+    @test size(counterfactual_data.X)[2] == 1000
+    @test size(counterfactual_data.X)[1] == 14
+    @test size(counterfactual_data.y)[2] == 1000
+
+    counterfactual_data = load_uci_adult(500)
+    @test size(counterfactual_data.X)[2] == 500
+    @test size(counterfactual_data.X)[1] == 14
+    @test size(counterfactual_data.y)[2] == 500
+    @test_throws ArgumentError load_uci_adult(0)
```
Comment (RaunoArike) [inline test/data.jl:75]: You could also add a test here that checks for an error when n is negative. Though this wouldn't increase code coverage, it would make the code more mutation-safe.

File: src/data/tabular/adult.jl [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+- `counterfactual_data::CounterfactualData`: A `CounterfactualData` object containing the preprocessed data.
+
+# Example
+data = load_uci_adult(20) # loads and preprocesses 20 samples from the Adult dataset
+"""
+function load_uci_adult(n::Union{Nothing,Int}=1000)
+    # Throw an exception if n < 1:
+    if !isnothing(n) && n < 1
+        throw(ArgumentError("n must be >= 1"))
+    end
```
Comment (RaunoArike) [inline src/data/tabular/adult.jl:19]: I think an error should also be thrown when n is bigger than the size of the dataset. For example, in the case of German credit data, an error is thrown when n > 1000.

File: test/data.jl (hunk 7 above)
Comment (RaunoArike) [reply test/data.jl:75]: Also, if you add a check that `n` is not bigger than the size of the dataset, then it would be good to add a test which checks that the error is indeed thrown when `n` is excessively big.

File: src/data/tabular/adult.jl (hunk 6 above)
Comment (A-Salmon) [reply src/data/tabular/adult.jl:50]: Right, good catch, slipped past me

File: test/data.jl (hunk 7 above)
Comment (A-Salmon) [reply test/data.jl:75]: Good point! I've added a check for n > 32000 (size of the dataset being 32561).

**Review by RaunoArike — APPROVED** [review]
Thanks for making the changes! Everything looks good now, approved.
